### PR TITLE
add Frontend build and Run Tests on CI/CD

### PR DESCRIPTION
**Pull request by yzhe819**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes/resolves # (issue) #60 

## Screenshots
Please include any screenshots or media that illustrates the changes made

## Type of change
Please delete options that are not relevant.

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
Leave blank if not applicable

I have completed these steps when making this pull request:
- [x] I have assigned my name to the issue
- [x] I have moved the issue to the **In Progress** column
- [x] I have labelled the PR appropriately
- [x] I have assigned myself to the PR

Before opening the PR for review:
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have moved the linked issue to the **Review in Progress** column

### CODE REVIEW

**Comment by frasermcc9** [comment]
Frontend should be a separate job imo. I think it should work like this:
- Setup job (setup node, install dependencies)
- Frontend job (waits for setup, builds and tests frontend)
- Backend job (waits for setup, builds and tests frontend)

This allows for the frontend and backend steps to run in parallel 

You can use the `needs` property to make a job wait on another.
https://docs.github.com/en/actions/using-workflows/advanced-workflow-features

**Comment by yzhe819** [comment]
> Cool, I will have a look. For the testing, I use the CodeQL to check the entire project. Is this a good idea? @justcrossheaven @frasermcc9

**Comment by j-chad** [comment]
I didn't know about codeql, very cool! You should probably put it in pr-check workflow though

**Comment by frasermcc9** [comment]
I was just meaning to run npm test but this can be added as well as long as it isn't going to make it impossible to merge things.

**Comment by justcrossheaven** [comment]
I think you can post its suggested changes first (if any and if significant) in the corresponding channel if you are adding it.

**Comment by yzhe819** [comment]
> I was just meaning to run npm test but this can be added as well as long as it isn't going to make it impossible to merge things.

OHH, I misunderstood I need to add extra front end analysis or tests. In order not to cause confusion, unnecessary codeQL will be removed.

**Comment by j-chad** [comment]
I think keep it. It already found a security hole in backend

**Comment by yzhe819** [comment]
> I didn't know about codeql, very cool! You should probably put it in pr-check workflow though

CodeQL is an interesting little tool. [CodeQL](https://codeql.github.com/)👏

**Comment by yzhe819** [comment]
> I think keep it. It already found a security hole in backend

Yes, that $eq one

**Comment by yzhe819** [comment]
This query injection has been fixed #73 
So, I can safely close this security alert. 
[Database query built from user-controlled sources](https://github.com/701-T4/platform/security/code-scanning/1)

More information can be found:
[whats-new-in-mongoose-6-sanitizefilter](https://thecodebarbarian.com/whats-new-in-mongoose-6-sanitizefilter.html)

File: .github/workflows/pr-check.yml
Comment (frasermcc9) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Build Frontend
```

File: .github/workflows/pr-check.yml
Comment (frasermcc9) [inline .github/workflows/pr-check.yml]: Can you add another step here (before the yarn test) that builds the backend? (wasn't in scope but imo should be done).

File: .github/workflows/pr-check.yml
Comment (frasermcc9) [inline .github/workflows/pr-check.yml:103]: Add a step for `yarn test` in the frontend

**Review by frasermcc9 — CHANGES_REQUESTED** [review]

**Comment by frasermcc9** [comment]
Think I'm happy to merge after those 3 things

File: .github/workflows/pr-check.yml
Comment (yzhe819) [reply .github/workflows/pr-check.yml:103]: cool, I will have a look

File: .github/workflows/pr-check.yml
Comment (yzhe819) [reply .github/workflows/pr-check.yml]: Thanks, I also add that one

**Comment by j-chad** [comment]
is this still supposed to be in draft?

**Comment by yzhe819** [comment]
> is this still supposed to be in draft?

ready for review💖

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]: Can we please just have 2 jobs: a `Backend` and `Frontend` job. Both of which should have a build and test step

**Review by j-chad — CHANGES_REQUESTED** [review]
Also please check the error in the last run. Other than these 2 things I'm very happy with it!

**Comment by j-chad** [comment]
also just a nit but analyze job doesn't need a matrix since we only have 1 language

File: .github/workflows/pr-check.yml
Comment (frasermcc9) [reply .github/workflows/pr-check.yml]: 3 jobs - 
`Setup` - install
`Backend` - build and test
`Frontend` - build and test

Comment (j-chad) [reply .github/workflows/pr-check.yml]: and `Analyse`

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml:60]: oh also please copy this if line to analyse job

**Review by j-chad — CHANGES_REQUESTED** [review]

**Comment by yzhe819** [comment]
> also just a nit but analyze job doesn't need a matrix since we only have 1 language

That's auto-generated documentation from git action marketplace, best left as it is.

**Comment by j-chad** [comment]
> That's auto-generated documentation from git action marketplace, best left as it is.

You can easily remove by removing the strategy on lines 23-28 and changing line 38 to `languages: javascript`. It will just make it a bit tidier on the overview screen

**Comment by yzhe819** [comment]
> > That's auto-generated documentation from git action marketplace, best left as it is.
> 
> You can easily remove by removing the strategy on lines 23-28 and changing line 38 to `languages: javascript`. It will just make it a bit tidier on the overview screen

It should be fine now. But still confused why yarn command fails.

**Comment by j-chad** [comment]
i think the working directory persists between steps. E.g. remove line 79 and 93

**Comment by yzhe819** [comment]
> i think the working directory persists between steps. E.g. remove line 79 and 93

No sure about that, if remove those working-directory `yarn build` will run as `lerna run build` under root directory.
Also, these errors occur in the first step.

File: .github/workflows/pr-check.yml
Comment (yzhe819) [reply .github/workflows/pr-check.yml]: > 3 jobs - `Setup` - install `Backend` - build and test `Frontend` - build and test

The latest one should be fine

**Comment by frasermcc9** [comment]
> i think the working directory persists between steps. E.g. remove line 79 and 93

it doesnt unless you set the defaults in the job iirc

**Comment by frasermcc9** [comment]
They should be `./packages` not `../`

**Comment by yzhe819** [comment]
> They should be `./packages` not `../`

`./packages` also not work, PR Check #249 used `./packages` but still failed

**Comment by frasermcc9** [comment]
> > They should be `./packages` not `../`
> 
> `./packages` also not work, PR Check #249 used `./packages` but still failed

Ok my bad. Turns out each job runs on a separate machine, so there's actually no point to the setup job. This means youll need to just copy paste the Setup parts before the `backend` and `frontend` jobs, but please when running the `yarn install` part, change the working-directory to only the relevant part (so we dont waste time doing frontend installs on backend etc)

**Comment by j-chad** [comment]
in this case I suggest we run the front end and back end tests within 1 job to save on setup time

**Comment by frasermcc9** [comment]
> in this case I suggest we run the front end and back end tests within 1 job to save on setup time

Doing it separately will be faster (avoid root level dependency installs) and run in parallel (faster), whilst only duplicating really quick tasks (setup node, checkout)

**Review by j-chad — CHANGES_REQUESTED** [review]
working directory needs to be set for the installation steps as well

**Comment by yzhe819** [comment]
> working directory needs to be set for the installation steps as well

**Comment by yzhe819** [comment]
> working directory needs to be set for the installation steps as well

Just a little confused. If we need installation for frontend and backend working directory separately. Do we need to run the `yarn install` under the root directory.

**Comment by frasermcc9** [comment]
> > working directory needs to be set for the installation steps as well
> 
> Just a little confused. If we need installation for frontend and backend working directory separately. Do we need to run the `yarn install` under the root directory.

no

**Comment by j-chad** [comment]
no we shouldn't need to

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Setup
```

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Test
```

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Build
```

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Setup
```

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Build
```

File: .github/workflows/pr-check.yml
Comment (j-chad) [inline .github/workflows/pr-check.yml]:
```suggestion
      - name: Test
```

**Review by j-chad — DISMISSED** [review]
LGTM! great work! just a few nitpicks, feel free to disregard

File: .github/workflows/pr-check.yml
Comment (yzhe819) [reply .github/workflows/pr-check.yml]: > LGTM! great work! just a few nitpicks, feel free to disregard

Cool!!!!! I just renamed some steps and waiting for @frasermcc9 's review. Once he's done reviewing, I can check the git action that are actually running.

**Review by frasermcc9 — DISMISSED** [review]

**Review by j-chad — DISMISSED** [review]
LGTM

File: .github/workflows/pr-check.yml
Comment (frasermcc9) [inline .github/workflows/pr-check.yml]: actually shouldnt this `yarn install` part be in another step (along with the other yarn install).

The actions aren't running which makes me think this is invalid.

Comment (frasermcc9) [reply .github/workflows/pr-check.yml]: Should be:
```yaml
      - name: Setup
        uses: actions/setup-node@v2
        with:
          node-version: '16'
          cache: 'yarn'

      - name: Install Dependencies
         working-directory: ./packages/frontend
         run: yarn install --prefer-offline --frozen-lockfile
```

and similar thing for backend shouldn't it?

Comment (yzhe819) [reply .github/workflows/pr-check.yml]: You are right, that one seems easier to understand.

**Comment by frasermcc9** [comment]
heh warnings treated as errors on CI

**Comment by yzhe819** [comment]
CI/CD should be fine now. @frasermcc9 and @j-chad Could you just check that frontend build error? And feel free to merge this pr.

**Comment by frasermcc9** [comment]
Yes this is all good now on the CICD end. I will fix the remaining issues.

**Review by frasermcc9 — APPROVED** [review]

**Review by j-chad — APPROVED** [review]
LGTM

**Pull request merged**
